src/main/java/fr/sncf/osrd/infra/signaling/AspectConstraint.java

```diff
@@ -0,0 +1,67 @@
+package fr.sncf.osrd.infra.signaling;
+
+import fr.sncf.osrd.train.Interaction;
+import fr.sncf.osrd.train.InteractionType;
+import fr.sncf.osrd.train.TrainState;
+import fr.sncf.osrd.train.phases.SignalNavigatePhase;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
+
+public abstract class AspectConstraint {
+
+    public static class ConstraintPosition {
+        public final double offset;
+        public final ConstraintPosition.Element element;
+
+        public enum Element {
+            CURRENT_SIGNAL,
+            NEXT_SIGNAL,
+            END
+        }
+
+        public ConstraintPosition(double offset, ConstraintPosition.Element element) {
+            this.offset = offset;
+            this.element = element;
+        }
+
+        /** Convert to a path position given a phase a train state */
+        public double convert(SignalNavigatePhase.State navigatePhase, TrainState trainState) {
+            Predicate<Interaction> isSignal = interaction -> interaction.interactionType == InteractionType.HEAD
+                    && interaction.actionPoint.getClass() == Signal.class;
+            Interaction interactionElement = null;
+            if (element == Element.NEXT_SIGNAL) {
+                AtomicBoolean isNext = new AtomicBoolean(false);
```

Comment:
Why do we need it to be atomic?

Reply:
To capture a variable it needs to be either an atomic or final.